File: Part7/Section 36 - Natural Language Processing/NLP.R
#Natural Lenguage Processing
#--->>Pre Procesado de Datos

#Importar Dataset

df_original=read.delim("Restaurant_Reviews.tsv",quote = '',stringsAsFactors = FALSE)

#----Limpieza del texto----
#install.packages("SnowballC")
#install.packages("tm")
library(tm)
library(SnowballC)
corpus=VCorpus(VectorSource(df_original$Review))
#convertir texto a solo mminusculas
corpus=tm_map(corpus,content_transformer(tolower))
#consltar resultado as.character(corpus[[1]])
#Eliminar Números
corpus=tm_map(corpus,removeNumbers)
#Eliminar Signos de Puntuación
corpus=tm_map(corpus,removePunctuation)
#Eliminar palabras no relevantes
corpus=tm_map(corpus,removeWords,stopwords(kind = "en"))
#Eliminar palabras derivadas
corpus=tm_map(corpus,stemDocument)
#Eliminar espacios extra
corpus=tm_map(corpus,stripWhitespace)

#Crear el modelo de Bag of Words
dtm=DocumentTermMatrix(corpus)
dtm=removeSparseTerms(dtm,0.999)

#----Aplicar Modelor de Clasificación----
#Random Forest

#Importar Dataset
df=as.data.frame(as.matrix(dtm))
df$Liked=df_original$Liked

# Codificar la variable de clasificación como factor
df$Liked = factor(df$Liked, levels = c(0,1))

# Dividir los datos en conjunto de entrenamiento y conjunto de test
# install.packages("caTools")
library(caTools)
set.seed(123)
split = sample.split(df$Liked, SplitRatio = 0.80)
training_set = subset(df, split == TRUE)
testing_set = subset(df, split == FALSE)

# Ajustar el Random Forest con el conjunto de entrenamiento.
#install.packages("randomForest")
library(randomForest)
classifier = randomForest(x = training_set[,-692],
                          y = training_set$Liked,
                          ntree = 10)

# Predicción de los resultados con el conjunto de testing
y_pred = predict(classifier, newdata = testing_set[,-692])

# Crear la matriz de confusión
cm = table(testing_set[, 692], y_pred)

File: Part7/Section 36 - Natural Language Processing/NLP.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Jan 31 19:52:41 2020

@author: root
"""


#Natural Lenguage Processing

#------<PREPROCESADO DE DATOS>----------

#Como importar una libreria

import  numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#importar dataset
df=pd.read_csv("Restaurant_Reviews.tsv",delimiter="\t",quoting=3)

#limpieza de texto
import re
import nltk
#nltk.download("stopwords")
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
corpus=[]
for i in range(0,1000):
    review=re.sub("[^a-zA-Z]"," ",
                  df['Review'][i])
    review=review.lower()
    review=review.split()
    ps=PorterStemmer()
    review=[ps.stem(word) for word in review if not word in set(stopwords.words("english"))]
    review=" ".join(review)
    corpus.append(review)
        
#Crear Bolsa de Palabras (Bag of Words)
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=1500)
X=cv.fit_transform(corpus).toarray()
y=df.iloc[:,-1].values


from sklearn.model_selection import train_test_split as splitter
x_train, x_test, y_train,  y_test = splitter(X, y, test_size=0.20, random_state=0)


#------<AJUSTAR MODELOS DE CLASIFICACION>----------

#Ajustar regresion con el el conjunto de entrenamiento

#Crear modelo de clasificación aqui 

from sklearn.naive_bayes import GaussianNB
classifier=GaussianNB()
classifier.fit(x_train,y_train)
print(classifier)


#predicción de nuestros modelos con conjunto de test

y_pred=classifier.predict(x_test)

#predicción de nuestros modelos con nuevo valor
#y_pred=clasifier.predict(6.5)

#crear matriz de confusión (evaluar prediccion)
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)
print(cm)